File: src/main/java/aoc2019/day09/Day9.java
package aoc2019.day09;

import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.Map;

import static aoc2019.day09.IntCodeProcesses.*;


public class Day9 {

    public static ArrayList<Long> readInput(String input) {
        ArrayList<Long> list = new ArrayList<>();
        BufferedReader reader;
        try {
            reader = new BufferedReader(new FileReader(input));
            String line = reader.readLine();
            while (line != null) {
                Arrays.stream(line.split(",")).map(Long::parseLong).forEachOrdered(list::add);
                line = reader.readLine();
            }
            reader.close();
        } catch (IOException e) {
            e.printStackTrace();
        }
        return list;
    }

    static long processInput(ArrayList<Long> numbers, int input) {
        int i = 0;
        int relativeBase = 0;
        StringBuilder outputBuilder = new StringBuilder();

        while (i < numbers.size()) {
            int opCode = Math.toIntExact(numbers.get(i));
            if (opCode == 99) {
                break;
            }

            Output output = processParameterMode(numbers, i, opCode, input, relativeBase);
            outputBuilder.append(output.getCode());

            if (output.getRelativeBase() != 0) {
                relativeBase = output.getRelativeBase();
            }
            i += output.getIndex();
        }

        return Long.parseLong(outputBuilder.toString());
    }

    public static Output processParameterMode(ArrayList<Long> numbers, int index, int opCode, int inputValue, int relativeBase) {
        int reducedOpCode = opCode % 100;
        Map<Integer, Integer> parameterModeMap = new HashMap<>();
        parameterModeMap.put(1, (opCode / 100) % 10);
        parameterModeMap.put(2, (opCode / 1000) % 10);
        parameterModeMap.put(3, (opCode / 10000) % 10);

        switch (reducedOpCode) {
            case 1:
            case 2:
                sumAndSubtractParam(numbers, index, reducedOpCode, relativeBase, parameterModeMap);
                return new Output("", 4);
            case 3:
            case 4:
                return inputAndOutputParam(numbers, index, reducedOpCode, relativeBase, parameterModeMap, inputValue);
            case 5:
            case 6:
                return new Output("", jumpIf(numbers, index, reducedOpCode, relativeBase, parameterModeMap));
            case 7:
            case 8:
                lessThanOrEquals(numbers, index, reducedOpCode, relativeBase, parameterModeMap);
                return new Output("", 4);
            case 9:
                return new Output("", 2, updateRelativeBase(numbers, index, relativeBase, parameterModeMap));
            default:
                return new Output("err");
        }
    }
}

File: src/main/kotlin/aoc2019/day09/Day9.kt
package aoc2019.day09

import java.io.BufferedReader
import java.io.FileReader
import java.io.IOException
import java.util.*


fun readInput(input: String): ArrayList<Long> {
    val list = ArrayList<Long>()
    val reader: BufferedReader
    try {
        reader = BufferedReader(FileReader(input))
        var line = reader.readLine()
        while (line != null) {
            Arrays.stream(line.split(",").toTypedArray()).map { s: String -> s.toLong() }
                .forEachOrdered { e: Long -> list.add(e) }
            line = reader.readLine()
        }
        reader.close()
    } catch (e: IOException) {
        e.printStackTrace()
    }
    return list
}


fun processInput(numbers: ArrayList<Long>, input: Int): Long {
    var i = 0
    var relativeBase = 0
    val outputBuilder = StringBuilder()
    while (i < numbers.size) {
        val opCode = Math.toIntExact(numbers[i])
        if (opCode == 99) {
            break
        }
        val output = processParameterMode(numbers, i, opCode, input, relativeBase)
        outputBuilder.append(output.code)
        if (output.relativeBase != 0) {
            relativeBase = output.relativeBase
        }
        i += output.index
    }
    return outputBuilder.toString().toLong()
}


fun processParameterMode(
    numbers: ArrayList<Long>,
    index: Int,
    opCode: Int,
    inputValue: Int,
    relativeBase: Int
): Output {
    val reducedOpCode = opCode % 100
    val parameterModeMap: MutableMap<Int, Int> = HashMap()
    parameterModeMap[1] = opCode / 100 % 10
    parameterModeMap[2] = opCode / 1000 % 10
    parameterModeMap[3] = opCode / 10000 % 10
    return when (reducedOpCode) {
        1, 2 -> {
            IntCodeProcesses.sumAndSubtractParam(
                numbers,
                index,
                reducedOpCode,
                relativeBase,
                parameterModeMap
            )
            Output("", 4)
        }
        3, 4 -> IntCodeProcesses.inputAndOutputParam(
            numbers,
            index,
            reducedOpCode,
            relativeBase,
            parameterModeMap,
            inputValue
        )
        5, 6 -> Output(
            "",
            IntCodeProcesses.jumpIf(numbers, index, reducedOpCode, relativeBase, parameterModeMap)
        )
        7, 8 -> {
            IntCodeProcesses.lessThanOrEquals(
                numbers,
                index,
                reducedOpCode,
                relativeBase,
                parameterModeMap
            )
            Output("", 4)
        }
        9 -> Output(
            "",
            2,
            IntCodeProcesses.updateRelativeBase(numbers, index, relativeBase, parameterModeMap)
        )
        else -> Output("err")
    }
}
